File: RB110/study/practice_problem_2/scramblies.rb
=begin


Complete the function scramble(str1, str2) that returns true if a portion of str1 characters can be rearranged 
to match str2, otherwise returns false.

Notes:

Only lower case letters will be used (a-z). No punctuation or digits will be included.
Performance needs to be considered.
Examples

scramble('rkqodlw', 'world') ==> True
scramble('cedewaraaossoqqyt', 'codewars') ==> True
scramble('katas', 'steak') ==> False

input:  
  -Two strings

Output:
  -True or False depending on weather a portion of rearranging string 1 will equalstring 2


Rules:
  Explicit:
    -Only lowercase alphabet letters will be used, no puncuation or digits

test cases exsamples:

p scramble('rkqodlw',           'world'     )#).to eq( true)
p scramble('cedewaraaossoqqyt', 'codewars'  )#).to eq( true)
p scramble('katas',             'steak'     )#).to eq(false)
p scramble('scriptjava',        'javascript')#).to eq( true)
p scramble('scriptingjava',     'javascript')#).to eq( true)

Data Structure:
  -strings
  -booleans

Asssiting structure:
  -Array

Algorithm:
  -Split str 2 into an array
  -iterate over that array
    -Check to see if each letter in array is included in string 1
      -Return false if not
  -Otherwise return true



=end


def scramble(str1, str2)
  str2.chars.each do |letter|
    return false if !str1.include?(letter)
  end
  true
end


p scramble('rkqodlw',           'world'     )#).to eq( true)
p scramble('cedewaraaossoqqyt', 'codewars'  )#).to eq( true)
p scramble('katas',             'steak'     )#).to eq(false)
p scramble('scriptjava',        'javascript')#).to eq( true)
p scramble('scriptingjava',     'javascript')#).to eq( true)

File: LS215/codewars/scramblies.js
/*
Complete the function scramble(str1, str2) that returns true if a portion of str1 characters can be rearranged 
to match str2, otherwise returns false.

Notes:

Only lower case letters will be used (a-z). No punctuation or digits will be included.
Performance needs to be considered.
Examples








*/

function scramble(str1, str2) {
  for (let i = 0; i < str2.length; i++ ) {
    if (!str1.includes(str2[i])) {
      return false;
    } else {
      str1 = str1.replace(str2[i], '');
    }
  }
  
  return true;
}


console.log(scramble('rkqodlw',           'world'      ))//, true );
console.log(scramble('cedewaraaossoqqyt', 'codewars'   ))//, true );
console.log(scramble('katas',             'steak'      ))//, false);
console.log(scramble('scriptjavx',        'javascript' ))//, false);
console.log(scramble('scriptingjava',     'javascript' ))//, true );
console.log(scramble('scriptsjava',       'javascripts'))//, true );
console.log(scramble('javscripts',        'javascript' ))//, false);
console.log(scramble('jscripts',          'javascript' ))//, false);
console.log(scramble('aabbcamaomsccdd',   'commas'     ))//, true );
console.log(scramble('commas',            'commas'     ))//, true );
console.log(scramble('sammoc',            'commas'     ))//, true )